### ISSUE TITLE
Use custom composer update/dump directory in updater

### ISSUE THREAD

**Issue by solverat** [pimcore/pimcore#2570]
We're using pimcore as a dependency, so we changed the `PIMCORE_COMPOSER_FILE_PATH` to the correct path for the pimcore directory (somewhere in `/vendor`), so `installComposerJson()` will update the correct file. This works smooth and does its job. But we have some real issues with the composer update procedure when using the pimcore web update command.

## Problem
There are two methods, [composerUpdate()](https://github.com/pimcore/pimcore/blob/c868dbe8ead074e141aa1f55a613b8bf49e8d21b/pimcore/lib/Pimcore/Update.php#L417) and [composerDumpAutoload()](https://github.com/pimcore/pimcore/blob/c868dbe8ead074e141aa1f55a613b8bf49e8d21b/pimcore/lib/Pimcore/Update.php#L448). Both of them force composer to use the `PIMCORE_COMPOSER_FILE_PATH` directory to interact with. If pimcore is running out of vendor, this is not the right place to do (also every dependency will be installed twice). 

## Solution
The solution is quite simple. We need another constant for the composer executable directory which will replace the  `-d PIMCORE_COMPOSER_FILE_PATH` in the composer process. For example `-d PIMCORE_COMPOSER_EXECUTABLE_DIRECTORY` With that we're able to update composer dependencies and the autoload dump within the right directory. 

## Additionals
For those who also implement pimcore as a dependency. You need to merge the pimcore composer.json with the merge plugin:

```
"extra": {
        "merge-plugin": {
            "include": [
                "./vendor/pimcore/pimcore/composer.json"
            ],
            "recurse": true,
            "replace": false,
            "merge-dev": true,
            "merge-extra": false,
            "merge-extra-deep": false,
            "merge-scripts": false
        }
    }
```

I just tested it with a hardcoded path and the update finished properly (i just tested it with the presta/sitemap-builder build. This bundle has been added after the update in the root vendor).

**Comment by solverat** [comment]
This what we would need to do in the `Update.php`:

```php
## here
public static function composerUpdate($options = [])
{
    $workingDir = defined('PIMCORE_COMPOSER_WORKING_DIR') ? PIMCORE_COMPOSER_WORKING_DIR : PIMCORE_COMPOSER_FILE_PATH;

    $composerLock = $workingDir . '/composer.lock';
     if (file_exists($composerLock)) {
         @unlink($composerLock);
    }

    # [..]

    $process = new Process($composerPath . ' update ' . implode(' ', $composerOptions) . ' -d ' . $workingDir); 

    # [..]
}

## and here
public static function composerDumpAutoload()
{
    $workingDir = defined('PIMCORE_COMPOSER_WORKING_DIR') ? PIMCORE_COMPOSER_WORKING_DIR : PIMCORE_COMPOSER_FILE_PATH;

    # [..]

    $process = new Process($composerPath . ' dumpautoload -d ' . $workingDir);

    # [..]

}

**Comment by solverat** [comment]
I just tested it further and there still a dependency issue. Example based on my previous comment: 

**root composer.json** 
```json
"require": {
  "pimcore/pimcore": "5.1.1"
}
```

**update pimcore via cli**: `bin/console pimcore:update -u5.1.3`

1. pimcore updates the `vendor/pimcore/pimcore/composer.json` 
2. pimcore triggers `composer update -d` (in main composer.json)

Since i'm including the pimcore `composer.json` from vendor via the `merge-plugin`, this will lead to a dependencies conflict:

**Message**: ` pimcore/pimcore v5.1.x requires pimcore/core-version v5.1.x -> satisfiable...`

That's because my root `composer.json` still has the `"pimcore/pimcore": "5.1.1"` entry, but the modified composer json of pimcore (which gets merged) requires the `"pimcore/core-version": "5.1.3"`

This is not that bad - simple adjust the root `composer.json` and start composer update again. But still - not perfect, especially if the composer update runs between multiple builds.

## Final Solution
Pimcore modifies the pimcore require tag in the root `composer.json` in `composerUpdate()`!

So instead of modifying the vendor composer (only if pimcore is a dependency, otherwise keep the current process) it would change the root composer from

```json
"require": {
  "pimcore/pimcore": "5.1.1"
}
```
to the current build. This build needs to add the current SHA to the require tag: 

```
"require": {
  "pimcore/pimcore": "5.1.3" # if current build is a version

  "pimcore/pimcore": "dev-master#2323332" # if current build only is a inbetween
}
```

After that, the composer update like described in my comment above will get triggered. With this, no composer merge is required and also no dependency issues.

**Comment by solverat** [comment]
We just updated several pimcore 5.x projects and it would be great if we would have this working dir CONST i described [here](https://github.com/pimcore/pimcore/issues/2570#issuecomment-363790158). We came to the conclusion that the more advanced option described in my second comment is more hacky than useful. As long pimcore is not a real dependency, we have to accept the situation as it is. :)

We also would provide a PR for that.
